paddle_image.py

```diff
@@ -0,0 +1,130 @@
+    ):
+        super().__init__(*args, **kwargs)
+        self.pool_strategy = pool_strategy
+        self.channel_axis = channel_axis
+        self.model_name = model_name
+        self._default_channel_axis = -3
+        self.inputs_name = None
+        self.outputs_name = None
+        self.on_gpu = on_gpu
+        self.default_batch_size = default_batch_size
+        self.default_traversal_paths = default_traversal_paths
+
+        import paddlehub as hub
+        module = hub.Module(name=self.model_name)
+        inputs, outputs, self.model = module.context(trainable=False)
+        self._get_inputs_and_outputs_name(inputs, outputs)
+
+        import paddle.fluid as fluid
+        self.device = fluid.CUDAPlace(0) if self.on_gpu else fluid.CPUPlace()
+        self.exe = fluid.Executor(self.device)
+
+    @requests
+    def encode(self, docs: DocumentArray, parameters: dict, **kwargs) -> DocumentArray:
+        if docs:
+            document_batches_generator = self._get_input_data(docs, parameters)
+            self._create_embeddings(document_batches_generator)
+
+    def _create_embeddings(self, document_batches_generator: Iterable):
+        for document_batch in document_batches_generator:
+            blob_batch = [d.blob for d in document_batch]
+            blob_batch = np.array(blob_batch)
+            if self.channel_axis != self._default_channel_axis:
+                blob_batch = np.moveaxis(blob_batch, self.channel_axis, self._default_channel_axis)
+            feature_map, *_ = self.exe.run(
+                program=self.model,
+                fetch_list=[self.outputs_name],
+                feed={self.inputs_name: blob_batch.astype('float32')},
+                return_numpy=True
+            )
+
+            if feature_map.ndim == 2 or self.pool_strategy is None:
+                embedding_batch = feature_map
+            else:
+                embedding_batch = self._get_pooling(feature_map)
+
+            for document, embedding in zip(document_batch, embedding_batch):
+                document.embedding = embedding
+
+    def _get_input_data(self, docs: DocumentArray, parameters: dict):
+        traversal_paths = parameters.get('traversal_paths', self.default_traversal_paths)
+        batch_size = parameters.get('batch_size', self.default_batch_size)
+
+        # traverse thought all documents which have to be processed
+        flat_docs = docs.traverse_flat(traversal_paths)
+
+        # filter out documents without images
+        filtered_docs = [doc for doc in flat_docs if doc.blob is not None]
+
+        return _batch_generator(filtered_docs, batch_size)
+
+    def _get_pooling(self, content: 'np.ndarray') -> 'np.ndarray':
+        """Get ndarray with selected pooling strategy"""
+        _reduce_axis = tuple((i for i in range(len(content.shape)) if i > 1))
+        return getattr(np, self.pool_strategy)(content, axis=_reduce_axis)
+
+    def _get_inputs_and_outputs_name(self, input_dict, output_dict):
```

Comment:
I would prefer a functional approach that sets `self.inputs_name` and `self.outputs_name` directly in the constructor, i.e. 
```
self.inputs_name, self.outputs_name = self._get_inputs_and_outputs_name(input, output)
```

Reply:
Yes, even more if u call it as `_getter`. It is much easier to read if the variables are explicitly initialized in `__init__`

Reply:
good point